snow/engine/snowman/straggler_detect.go

```diff
@@ -0,0 +1,331 @@
+	stakeThresholdForStragglerSuspicion         = 0.75
+	minimumStakeThresholdRequiredForNetworkInfo = 0.8
+	knownStakeThresholdRequiredForAnalysis      = 0.8
+)
+
+type stragglerDetectorConfig struct {
+	// getTime returns the current time
+	getTime func() time.Time
+
+	// minStragglerCheckInterval determines how frequently we are allowed to check if we are stragglers.
+	minStragglerCheckInterval time.Duration
+
+	// log logs events
+	log logging.Logger
+
+	// minConfirmationThreshold is the minimum stake percentage that below it, we do not check if we are stragglers.
+	minConfirmationThreshold float64
+
+	// connectedPercent returns the stake percentage of connected nodes.
+	connectedPercent func() float64
+
+	// connectedValidators returns a set of tuples of NodeID and corresponding weight.
+	connectedValidators func() set.Set[ids.NodeWeight]
+
+	// lastAcceptedByNodeID returns the last accepted height a node has reported, or false if it is unknown.
+	lastAcceptedByNodeID func(id ids.NodeID) (ids.ID, bool)
+
+	// processing returns whether this block ID is known and its descendants have not yet been accepted by consensus.
+	// This means that when the last accepted block is given as input, true is returned, as by definition
+	// its descendants have not been accepted by consensus, but this block is known.
+	// For any block ID belonging to an ancestor of the last accepted block, false is returned,
+	// as the last accepted block has been accepted by consensus.
+	processing func(id ids.ID) bool
+
+	// lastAccepted returns the last accepted block of this node.
+	lastAccepted func() ids.ID
+
+	// getSnapshot returns a snapshot of the network's nodes and their last accepted blocks,
+	// or false if it fails from some reason.
+	getSnapshot func() (snapshot, bool)
+
+	// haveWeFailedCatchingUp returns whether we have not replicated enough blocks of the given snapshot
+	haveWeFailedCatchingUp func(snapshot) bool
+}
+
+type stragglerDetector struct {
+	stragglerDetectorConfig
+
+	// continuousStragglingPeriod defines the time we have been straggling continuously.
+	continuousStragglingPeriod time.Duration
+
+	// previousStragglerCheckTime is the last time we checked whether
+	// our block height is behind the rest of the network
+	previousStragglerCheckTime time.Time
+
+	// prevSnapshot is the snapshot from a past iteration.
+	prevSnapshot snapshot
+}
+
+func newStragglerDetector(
+	getTime func() time.Time,
+	log logging.Logger,
+	minConfirmationThreshold float64,
+	lastAccepted func() (ids.ID, uint64),
+	connectedValidators func() set.Set[ids.NodeWeight],
+	connectedPercent func() float64,
+	processing func(id ids.ID) bool,
+	lastAcceptedByNodeID func(ids.NodeID) (ids.ID, bool),
+) *stragglerDetector {
+	sd := &stragglerDetector{
+		stragglerDetectorConfig: stragglerDetectorConfig{
+			lastAccepted:              dropHeight(lastAccepted),
+			processing:                processing,
+			minStragglerCheckInterval: minStragglerCheckInterval,
```

Comment:
Yes, because it would make sense to add a config parameter to override it, and also for testing it would make sense to have it configurable.

Reply:
I agree with having this as a configurable param. but I'm not sure how much we can deem this as a configurable if we don't export this config + not take the config struct as a param to the function. (this could be a good CLI flag to enable/disable this altogether)

Reply:
In the latest version of this code, this field is part of `stragglerDetectorConfig` which is initialized in `engine_decorator.go`, and is passed into the constructor. 

Therefore, this is configurable at the level of this PR. 

Making it configurable and rippling it down from the CLI flags should be taken care of at a later stage (note that I marked this PR as not to be merged just yet)